### ISSUE TITLE
Instrument for PhET-iO

### ISSUE THREAD

**Issue by samreid** [phetsims/molarity#31]
@ariel-phet and I would like to reserve the Molarity simulation for @pixelzoom to instrument for PhET-iO.  It is on the simpler side of simulations and would be an ideal candidate for @pixelzoom to take the "How to Instrument a PhET Simulation for PhET-iO" instructions https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md for a test drive.

This issue is mainly a placeholder so that nobody else tries to instrument the simulation first, and to put it on @pixelzoom's radar, hence the "on-hold" label.  After Unit Rates is published, we can determine a schedule for instrumenting this simulation for PhET-iO.  I would recommend scheduling about 2 weeks for this work, not because it is a complex simulation but because it will likely lead to many valuable questions and discussions about process and strategy.

**Comment by pixelzoom** [comment]
Self assigning so that it's on my radar.  

Based on current milestones for Unit Rates, followed by vacation, it's unlikely that I will get to this until April.

**Comment by ariel-phet** [comment]
@pixelzoom sounds good.

**Comment by pixelzoom** [comment]
Another sim has been prioritize ahead of this, see https://github.com/phetsims/function-builder/issues/103.  That will likely push this out until sometime in mid-late May.

**Comment by pixelzoom** [comment]
Checklist, from https://github.com/phetsims/phet-io/edit/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md:

 - [ ] Copy this checklist to the issue description (top issue comment) for tracking.
 - [ ] [Understand the goal](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#understand-the-goal)
 - [ ] [Code review](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#code-review)
 - [ ] Use Property instead of PropertySet
 - [ ] [Import Tandem to main.js](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#import-tandem-to-mainjs)
 - [ ] [Create root tandem](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#create-root-tandem)
 - [ ] [Pass tandems to sim screens](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#pass-tandems-to-sim-screens)
 - [ ] [Pass tandems to model and view](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#pass-tandems-to-model-and-view)
 - [ ] [Pass tandems to all constructors](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#pass-tandems-to-all-constructors)
   + See also [Testing while adding Tandem instances](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#testing-while-adding-tandem-instances). Test with `?phetioValidateTandems=false` since common code components don't yet have tandems.
 - [ ] [Pass tandem instances to common code components](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#pass-tandem-instances-to-common-code-components)
   + See also [Testing while adding Tandem instances](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#testing-while-adding-tandem-instances).
 - [ ] [Replace types with tandemized equivalents](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#replace-types-with-tandemized-equivalents)
 - [ ] [Instrument any Uninstrumented Common Code Components Used by the Sim](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#instrument-any-uninstrumented-common-code-components-used-by-the-sim)
 - [ ] [Instrument new component types defined by the sim](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#instrument-new-component-types-defined-by-the-sim)
 - [ ] [Add simulation specific support](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#add-simulation-specific-support)
 - [ ] [Support dynamic state](https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md#support-dynamic-state)
 - [ ] Make sure Joist `dt` values are used instead of Date.now() or other Date functions.  This is necessary for reproducible playback via input events.
 - [ ] Are random numbers using `phet.joist.random`, and all doing so after modules are declared (non-statically)?  For
 instance, the following methods (and perhaps others) should not be used: `Math.random`, `_.shuffle`, `_.sample`, `_.random`
 - [ ] `undefined` values cannot be saved from the phet-io state save, sims should be written to use `null` instead
 - [ ] Verify that the simulation works in all of the phet-io wrappers
 - [ ] Set up a task to have the newly instrumented sim tested by the QA team
 - [ ] Update these instructions if you find them to be incomplete, inconsistent or incorrect
 - [ ] Add your instrumented simulation to the spreadsheet here: https://docs.google.com/spreadsheets/d/1ToW8wc9zOFpMglIf63jB1j8YA9Qmh2dxNLH2Sm4qxEo/edit?usp=drive_web

**Comment by pixelzoom** [comment]
Creating this comment to collect questions and issues encountered while instrumenting. Some of these may be broken out into separate GitHub issues.

- [ ] It's apparently not possible to incrementally test as you go. After completing the "Pass tandems to sim screens" step, I was unable to test with "?brand=phet-io&phetioLog=lines&phetioStandalone&ea&phetioValidateTandems=false", due to:
![screenshot_201](https://cloud.githubusercontent.com/assets/3046552/24819983/1b15efbe-1ba4-11e7-868f-c18f259fed8e.png)

-[ ] In step "Pass tandems to model and view", it was unclear whether ScreenView subtypes should propagate their tandem to ScreenView.  Looking at other examples, the answer appears to be "no".  But this begs the question of how to know when to propagate.

**Comment by samreid** [comment]
> What are the things with molarity.ariaHerald prefix?

Please see the documentation at the top of scenery-phet/js/accessibility/AriaHerald.js

**Comment by samreid** [comment]
> The current rule is "if any ancestor in the type hierarchy supports tandem, you should propagate the tandem”. But that's difficult to determine unless you look through the entire type hierarchy.

I agree with the rule as you stated it and in the difficulty of looking through the type hierarchy.  However, this is not too different than our normal procedure for specifying options.  If C extends B extends A and I want to know if I can pass a backgroundColor option when I call `new C` then in the worst case scenario I have to look in C and B and A.  It seems the same problem applies to optional tandems.  

However, ever since https://github.com/phetsims/phet-io/issues/926 was completed, all Nodes support Tandem.  So if you know that your subclass is a Node then you can pass a tandem into the superclass without having to look through N files.  With model classes you have to look through the inheritance hierarchy, but there is a lot more common code for views than for models.

**Comment by pixelzoom** [comment]
I tried all of the controls in the instance-proxies wrapper:

http://localhost/~cmalley/GitHub/phet-io/wrappers/instance-proxies/instance-proxies.html?sim=molarity&ea

The only one that didn't appear to do anything was this tandem, which looks like it might be a bug in sun.ComboBox:

`molarity.molarityScreen.view.soluteComboBox.buttonNode.itemNode`

**Comment by pixelzoom** [comment]
Regarding test and publication, here's my Skype thread with @ariel-phet:

[4/11/17, 2:47:11 PM] Chris Malley: Molarity is instrumented for PhET-iO, and SR says that it should be published.  So it will require dev and RC test.  Do you also want to publish a public (phet brand) version at the same time?  Seems like yes, since phet brand testing is a subset of phetio brand testing.
[4/11/17, 2:48:16 PM] Ariel Paul: Yes, seems reasonable to retest and publish the public one, picking up some recent features

So now we have a problem.  phet and phetio brands are not typically published at the same time, and the current process makes it difficult/impossible to do so.   For example, phetio versions have their own release branch naming scheme, e.g. 1.1-phetio.   

I'm going to hold off on this until we straighten out the process and address the version id issue (https://github.com/phetsims/chipper/issues/560).

**Comment by samreid** [comment]
> the current process makes it difficult/impossible to do so

Perhaps the workaround given the pre-existing process would be to create "1.3" and "1.3-phetio" branches, and maintain both.  We can discuss alternatives in the near future, but this seems a safe way to proceed at the moment.

> I'm going to hold off on this until we straighten out the process and address the version id issue (phetsims/chipper#560).

I recommend not making molarity work contingent on version name changes.  Even if we agree to change the versions as recommended in phetsims/chipper#560 it will likely be several weeks until that can be implemented.  The existing versioning scheme has worked acceptably until now and I don't think changing it is a good reason to postpone moving forward with Molarity.

**Comment by pixelzoom** [comment]
OK, let's suppose that I follow the above suggestion.

Will there be separate test tasks, or 1 test task?
Will there be separate test matrices or 1 matrix?
If there's 1 task and matrix, what URLs are tested?

**Comment by samreid** [comment]
Great questions, we are in uncharted waters since we haven't done "phet" at the same time as "phet-io" ever before.  Perhaps @phet-steele can recommend how to set this up?

**Comment by pixelzoom** [comment]
Labeling for developer meeting to see if any other developers have suggestions about how to proceed with testing and publishing 'phet' and 'phetio' brands simultaneously.

**Comment by jonathanolson** [comment]
Presumably there would be two separate branches, one for each brand?

**Comment by pixelzoom** [comment]
Dev meeting decision:

• I will create separate branches with the same shas.
• I will create separate test tasks.
• @phet-steele will create separate test matrices
• @phet-steele will test phet brand first, testing all sim features
• @phet-steele will test phet-io brand second, testing only phet-io features

**Comment by pixelzoom** [comment]
Dev testing has begun in https://github.com/phetsims/tasks/issues/819.

**Comment by pixelzoom** [comment]
Dev testing has completed in https://github.com/phetsims/tasks/issues/865.  On to RC testing...

**Comment by pixelzoom** [comment]
RC testing is completed (#13,#14).  I'll proceed to publication (#38,#39).

**Comment by pixelzoom** [comment]
The phet and phet-io version have been published, and I've completed the checklist in https://github.com/phetsims/molarity/issues/31#issuecomment-292646994.

**Comment by pixelzoom** [comment]
The original purpose of this issue was to gain experience with instrumenting and publishing a sim. I've provided my feedback in the comments above, and in https://github.com/phetsims/phet-io/issues/1137. 

In summary:

Pros:
• Instrumenting a sim is relatively straightforward (if there's a good match between PhET-iO requirements and how the sim was implemented).
• The API provides control over just about everything.

Cons:
• The API has an almost total lack of abstraction/encapsulation, and that will cause longterm headaches for PhET and customers. 
• Instrumentation can create memory leaks, often subtle.
• Instrumenting everything is overkill. It requires testing and debugging things that customers will never use, and allows customers to do things that PhET likely doesn't want them to do. 
• If there's not a good match between PhET-iO requirements and how a sim was implemented, then instrumentation can require rewriting parts of the sim.
• The processes and tools to support building and deploying sims needs a lot of work.

@ariel-phet Anything else you'd like from this issue?

**Comment by pixelzoom** [comment]
@ariel-phet @samreid Anything else you'd like from this issue?

**Comment by pixelzoom** [comment]
@ariel-phet @samreid ping.

**Comment by samreid** [comment]
I'm all set, thanks for your work @pixelzoom!  I'll create a separate issue to remind @kathy-phet that she can share this to our partners.
